File: solutions/487. Max Consecutive Ones II.py
class Solution:
    def findMaxConsecutiveOnes(self, nums: List[int]) -> int: 
        # 1004. Max Consecutive Ones III, runtime O(N), space O(1)
        k, ans = -1, 0
        deq = deque([])
        for i in range(len(nums)):
            if nums[i] == 0:
                deq.append(i)
            if len(deq) > 1:
                k = deq.popleft()
            ans = max(ans, i-k)
        return ans
        
#         # two pointer, Runtime O(N), space O(1)
#         k = -1 
#         j = 0 # last seen 0 position
#         cnt_zero = 0
#         ans = 0
        
#         for i in range(len(nums)):
#             if nums[i] == 0:
#                 cnt_zero += 1
#                 if cnt_zero > 1:
#                     k = j
#                 j = i
#             ans = max(ans, i-k)
#         return ans
    
    #  (i, k, j, ans)
    # [1,0,1,1,0] 
    # (0, -1, 0, 1) -> (1, -1, 1, 2) -> (2,-1,1,3) -> (3,-1,1,4) -> (4,1,4,4)

File: solutions/487. Max Consecutive Ones II.cpp
class Solution {
public:
    int findMaxConsecutiveOnes(vector<int>& nums) {
        // 2nd attempt, sliding window
        int ans=0, l=0, cnt=0;
        
        for (int i=0; i<nums.size(); i++) {
            if (nums[i] == 0) cnt+=1;
        
        while (cnt==2) {
            if (nums[l]==0) {
                cnt-=1;
            }
            l+=1;
        }
        ans = max(ans, i-l+1);
        }
        return ans;
        
        // First attempt, two-pointer, time O(N), space O(1)
        // int ans=0, cnt=0, zeroLastSeen=-1, zeroPos=-1;
        // for (int i=0; i<nums.size(); i++) {
        //     if (nums[i]==0) {
        //         if (cnt>0) {
        //             zeroLastSeen = zeroPos;
        //             zeroPos = i;
        //         }
        //         else {
        //             cnt += 1;
        //             zeroPos=i;
        //         }
        //     }
        //     ans = max(ans, i-zeroLastSeen);
        // }
        // return ans;
    }
};
